### PR TITLE
Fix #8650 for improved hidden files finder performance

### LINKED ISSUE #8650: Faster Find of hidden files
In profiling the delay between rubocop startup and the beginning of scanning files I discovered that rubocop will traverse the project directories/files twice.  Once for `all_files` and then again to find `hidden_files`.

```ruby
# lib/rubocop/target_finder.rb
def target_files_in_dir(base_dir = Dir.pwd)
  # ...
  all_files = find_files(base_dir, File::FNM_DOTMATCH)
  hidden_files = Set.new(all_files - find_files(base_dir, 0))
  #...
end

#...
def find_files(base_dir, flags)
  # ...
  Dir.glob(pattern, flags).select { |path| FileTest.file?(path) }
end
```

In the code above, `#find_files` is called twice which calls Dir.glob.  Each time with a different flag.  This is ineficient as it's touching the local disk twice for ever non-hidden file and it's checking every entry0for `FileTest.file?(path)` twice . 

## Describe the solution you'd like

A more efficient solution would be to only call `#find_files` once and then filter for hidden files.

```diff
- hidden_files = Set.new(all_files - find_files(base_dir, 0))
+ hidden_files = all_files.select {|file|file =~ /[\/|\\]\./ }
```

The regex looks for any file paths which include a '/.' (Linux) or ''\\.'' (Windows)

## Benchmarks

I tested this code change in a medium size Rails project.

### Project Stats Stats:

```
all_files: 123,727
hidden_files: 2,553
target_files: 405
```
Note: The majority of the files are in engines/my-engine/node_modules directory (which is a separate issue).

```ruby
require "benchmark/ips"

Benchmark.ips(7) do |x|
  x.report("glob:") { hidden_files = Set.new(all_files - find_files(base_dir, 0)) }
  x.report("select:") { hidden_files = all_files.select {|file|file =~ /\/\./ } }
  x.compare!
end
```

### Results:

```
Comparison:
              select:       23.6 i/s
                glob:        1.0 i/s - 23.99x  (± 0.00) slower

```

The above IPS benchmark shows that on this project the 'select' code can ran 23 times within 1 second, whereas the glob could only run once within a second.  In other words, glob is 23.99x slower.

## Additional context

If this code change is desired, I can create an MR.

### PR DESCRIPTION
**Pull request by tleish**
Faster find of hidden files in `TargetFinder` class which improves rubocop initial startup speed.

Benchmarks now show a 57x improvement in one large project tested:
```
Comparison:
             select:       50.6 i/s
               glob:        0.9 i/s - 57.38x  (± 0.00) slower
```

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [ ] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by marcandre** [comment]
57x is nice 😆
Could you not use `grep` with a regexp here?

**Comment by tleish** [comment]
> Could you not use `grep` with a regexp here?

@marcandre - I previously tested this, but turns out that the proposed solution is faster.

```ruby
      require "benchmark/ips"

      Benchmark.ips(7) do |x|
        x.report("glob") { hidden_files = Set.new(all_files - find_files(base_dir, 0)) }
        x.report("select") { hidden_files = all_files.select { |file| file.include?(HIDDEN_FILE_PATTERN) } }
        x.report("grep") { hidden_files = all_files.grep(%r{#{File::SEPARATOR}\.}) }
        x.compare!
      end
```

```
Comparison:
             select:       52.3 i/s
               grep:       23.5 i/s - 2.22x  (± 0.00) slower
               glob:        0.9 i/s - 59.16x  (± 0.00) slower
```

**Comment by marcandre** [comment]
> > Could you not use `grep` with a regexp here?
> 
> @marcandre - I previously tested this, but turns out that the proposed solution is faster.

Ah, right... Forgot about https://bugs.ruby-lang.org/issues/17030

**Comment by tleish** [comment]
> Ah, right... Forgot about https://bugs.ruby-lang.org/issues/17030

FYI

```ruby
Benchmark.ips(7) do |x|
    x.report("glob") { hidden_files = Set.new(all_files - find_files(base_dir, 0)) }
    x.report("include?") { hidden_files = all_files.select { |file| file.include?(HIDDEN_FILE_PATTERN) } }
    x.report("=~") { hidden_files = all_files.select { |file| file =~ grep_hidden_files_pattern } }
    x.report("match?") { hidden_files = all_files.select { |file| file.match?(grep_hidden_files_pattern) } }
    x.report("grep") { hidden_files = all_files.grep(grep_hidden_files_pattern) }
    x.compare!
end
```

```
Comparison:
            include?:       50.2 i/s
              match?:       42.9 i/s - 1.17x  (± 0.00) slower
                grep:       24.6 i/s - 2.04x  (± 0.00) slower
                  =~:       22.9 i/s - 2.19x  (± 0.00) slower
                glob:        0.9 i/s - 56.91x  (± 0.00) slower
```

File: lib/rubocop/target_finder.rb [hunk 1]
```diff
@@ -5,6 +5,8 @@ module RuboCop
   # and picking ruby files.
   # @api private
   class TargetFinder
+    HIDDEN_FILE_PATTERN = "#{File::SEPARATOR}."
```
Comment (bbatsov) [inline lib/rubocop/target_finder.rb:8]: I'm not sure that naming this "pattern" is a good idea, as it makes me think of regular expressions. :-)

File: lib/rubocop/target_finder.rb [hunk 2]
```diff
@@ -55,7 +57,7 @@ def target_files_in_dir(base_dir = Dir.pwd)
       # Support Windows: Backslashes from command-line -> forward slashes
       base_dir = base_dir.gsub(File::ALT_SEPARATOR, File::SEPARATOR) if File::ALT_SEPARATOR
       all_files = find_files(base_dir, File::FNM_DOTMATCH)
-      hidden_files = Set.new(all_files - find_files(base_dir, 0))
+      hidden_files = all_files.select { |file| file.include?(HIDDEN_FILE_PATTERN) }
```
Comment (bbatsov) [inline lib/rubocop/target_finder.rb:60]: I'd add a note we're doing this for performance reasons, as someone might want to optimize it for readability or whatever.

File: lib/rubocop/target_finder.rb (hunk 1 above)
Comment (tleish) [reply lib/rubocop/target_finder.rb]: updated to `HIDDEN_PATH_SUBSTRING`

File: lib/rubocop/target_finder.rb (hunk 2 above)
Comment (tleish) [reply lib/rubocop/target_finder.rb]: adde comment

**Pull request merged**